src/main/groovy/wooga/gradle/appcenter/api/AppCenterReleaseUploader.groovy

```diff
@@ -0,0 +1,425 @@
+        String releaseId
+        ReleaseUploadStatus uploadStatus
+        String errorDetails
+        String releaseDistinctId
+        URI releaseUrl
+
+        private ReleaseUpload(String releaseId, ReleaseUploadStatus uploadStatus, String errorDetails, String releaseDistinctId, URI releaseUrl) {
+            this.releaseId = releaseId
+            this.uploadStatus = uploadStatus
+            this.errorDetails = errorDetails
+            this.releaseDistinctId = releaseDistinctId
+            this.releaseUrl = releaseUrl
+        }
+
+        static ReleaseUpload fromResponse(HttpResponse response) {
+            def responseBody = getResponseBody(response)
+            String releaseId = responseBody['id']
+            ReleaseUploadStatus uploadStatus = responseBody['upload_status'] as ReleaseUploadStatus
+            String releaseDistinctId = null
+            URI releaseUrl = null
+            String errorDetails = null
+
+            switch (uploadStatus) {
+                case ReleaseUploadStatus.readyToBePublished:
+                    releaseDistinctId = responseBody['release_distinct_id']
+                    releaseUrl = new URI(responseBody['release_url'].toString())
+                    break
+                case ReleaseUploadStatus.error:
+                    errorDetails = responseBody['error_details']
+                    break
+            }
+
+            new ReleaseUpload(releaseId, uploadStatus, errorDetails, releaseDistinctId, releaseUrl)
+        }
+    }
+
+    static Logger logger = AppCenterRest.logger
+    static API_BASE_URL = AppCenterRest.API_BASE_URL
+
+    final HttpClient client
+    final String owner
+    final String applicationIdentifier
+    final String apiToken
+
+    AppCenterReleaseUploader(HttpClient client,
+                             String owner,
+                             String applicationIdentifier,
+                             String apiToken) {
+        this.client = client
+        this.owner = owner
+        this.applicationIdentifier = applicationIdentifier
+        this.apiToken = apiToken
+    }
+
+    void Upload(UploadArguments args) {
+
+        CreateReleaseUpload releaseUpload = createReleaseUpload(args.buildVersion, args.buildNumber)
+        uploadFile(releaseUpload, args.binary)
+        updateReleaseUpload(releaseUpload.id, ReleaseUploadStatus.uploadFinished)
+
+        String releaseId = pollForReleaseId(releaseUpload.id)
+        Map release = getRelease(releaseId)
+        String downloadUrl = release["download_url"].toString()
+        String installUrl = release["install_url"].toString()
+
+        distribute(releaseId, args.destinations, args.releaseNotes, args.buildInfo)
+        logger.info("published to AppCenter release: ${releaseId}")
+        logger.info("download_url: ${downloadUrl}")
+        logger.info("install_url: ${installUrl}")
+
+        args.uploadVersionMetaData.asFile << JsonOutput.prettyPrint(JsonOutput.toJson(release))
```

Comment:
This should move back into the gradle task. Upload should then return the `Map` object or maybe a parsed result. I'm fine with the map at the moment. This also means that uploadVersionMetaData is not part of this uplaoder. Sorry I overlooked this property.

Reply:
Understood.